spec/src/main/asciidoc/architecture.adoc

```diff
@@ -97,6 +97,28 @@ Tags have taken over the role to, for example, identify an application (`app=myS
 (`tier=database` or `tier=app_server`) and also the node/container id. Aggregation of metrics can then work over label
 queries (Give me the API hit count for `app=myShop && tier=app_server`).
 
+In MicroProfile Metrics, tags add an additional dimension to metrics that share a common basis. For example, a metric named
+`carCount` can be further differentiated by the car type (sedan, SUV, coupe, and etc) and the colour (red, blue, white, black,
+and etc). Rather than incorporating this in the metric name, tags can be used to capture this information in separate metrics.
+
+[source]
+----
+carCount{type=sedan,colour=red}
+carCount{type=sedan,colour=blue}
+carCount{type=suv,colour=red}
+carCount{type=coupe,colour=blue}
+----
+
+For portability reasons, the key name for the tag must match the regex `[a-zA-Z_][a-zA-Z0-9_]*` (Ascii alphabet, numbers and underscore).
+If an illegal character is used, the implementation must throw an IllegalArgumentException.
+If a duplicate tag is used, the last occurrence of the tag is used.
+
+The tag value may contain any UTF-8 encoded character.
```

Comment:
Looking at how tags are being parsed in `MetricID` class, I suppose the values may not contain `,` and `=`, otherwise they will not be parsed properly. So either the parsing is wrong and we need to employ some mechanism for escaping these characters, or we have to forbid using `,` and `=` in the values.

Reply:
The `=` should be fine, keys are alphanumeric so we take the first equal sign and everything before is a key and everything after is the value.
However, yes, the `,` is definitely problematic!

We can change the`MetricRegistry.register(Metadata metadata, T metric, String tags)`, `MetricRegistry.counted(Metadata metadata, String tags)`, etc... to
`MetricRegistry.register(Metadata metadata, T metric, String... tags)`, `MetricRegistry.counted(Metadata metadata, String... tags)`, etc where tags are individual "key=value" at which point we'd need to feed in the name and a String Array to MetricsID instead.

OR have Tags object that you build up with `.addtag(String kvpair)` and/or `.addtag(String key, String value)` and you feed in to
`MetricRegistry.register(Metadata metadata, T metric, Tags tags)`,` MetricRegistry.counted(Metadata metadata, Tags tags)`, etc and then we'd feed the name and Tag into MetricsID instead

One thing to consider is the MP_METRIC_TAGS env variable where it's one whole string. There we might need to set the rule of key="value",key="value" and any intentional quotes inside the value will need to be escaped?

Reply:
Yeah the MP_METRIC_TAGS variable will be a bit of a problem for the `=` sign. Do we really need to allow the `=` sign? I mean, without it, the parsing will be easier, and everything will also be easier and less error-prone for the user. Will real world use cases need K=V pairs like `key="otherkey=value"`? If it is deemed necessary, then yeah the quoting you are suggesting is probably the way to go.

Other than that, I am for `.addTag(String key, String value)` method, it feels like Java, and it's type-safe and easy, definitely better than accepting a String containing `key=value` (or even `key1=value1`,`key2=value2`) and having to parse it.

Reply:
Right and even also having `,` as a valid value character would necessitate having quotes around the value. Also should also consider @donbourne suggestion as proposal #3 on the gitter page https://gitter.im/eclipse/microprofile-metrics